pkg/core/dao.go

```diff
@@ -0,0 +1,601 @@
+	if err != nil {
+		return nil, 0, err
+	}
+	r := io.NewBinReaderFromBuf(b)
+
+	var height uint32
+	r.ReadLE(&height)
+
+	tx := &transaction.Transaction{}
+	tx.DecodeBinary(r)
+	if r.Err != nil {
+		return nil, 0, r.Err
+	}
+
+	return tx, height, nil
+}
+
+// PutVersion stores the given version in the underlying Store.
+func (dao *Dao) PutVersion(v string) error {
+	return dao.store.Put(storage.SYSVersion.Bytes(), []byte(v))
+}
+
+func (dao *Dao) PutCurrentHeader(hashAndIndex []byte) error {
+	return dao.store.Put(storage.SYSCurrentHeader.Bytes(), hashAndIndex)
+}
+
+// read2000Uint256Hashes attempts to read 2000 Uint256 hashes from
+// the given byte array.
+func read2000Uint256Hashes(b []byte) ([]util.Uint256, error) {
+	r := bytes.NewReader(b)
+	br := io.NewBinReaderFromIO(r)
+	lenHashes := br.ReadVarUint()
+	hashes := make([]util.Uint256, lenHashes)
+	br.ReadLE(hashes)
+	if br.Err != nil {
+		return nil, br.Err
+	}
+	return hashes, nil
+}
+
+// checkTransactionInStore returns true if the given store contains the given
+// Transaction hash.
+func (dao *Dao) HasTransaction(hash util.Uint256) bool {
+	key := storage.AppendPrefix(storage.DataTransaction, hash.BytesReverse())
+	if _, err := dao.store.Get(key); err == nil {
+		return true
+	}
+	return false
+}
+
+// StoreAsBlock stores the given block as DataBlock.
+func (dao *Dao) StoreAsBlock(block *Block, sysFee uint32) error {
+	var (
+		key = storage.AppendPrefix(storage.DataBlock, block.Hash().BytesReverse())
+		buf = io.NewBufBinWriter()
+	)
+	// sysFee needs to be handled somehow
+	//	buf.WriteLE(sysFee)
+	b, err := block.Trim()
+	if err != nil {
+		return err
+	}
+	buf.WriteLE(b)
+	if buf.Err != nil {
+		return buf.Err
+	}
+	return dao.store.Put(key, buf.Bytes())
+}
+
+// StoreAsCurrentBlock stores the given block witch prefix SYSCurrentBlock.
+func (dao *Dao) StoreAsCurrentBlock(block *Block) error {
+	buf := io.NewBufBinWriter()
+	buf.WriteLE(block.Hash().BytesReverse())
+	buf.WriteLE(block.Index)
+	return dao.store.Put(storage.SYSCurrentBlock.Bytes(), buf.Bytes())
+}
+
+// StoreAsTransaction stores the given TX as DataTransaction.
+func (dao *Dao) StoreAsTransaction(tx *transaction.Transaction, index uint32) error {
+	key := storage.AppendPrefix(storage.DataTransaction, tx.Hash().BytesReverse())
+	buf := io.NewBufBinWriter()
+	buf.WriteLE(index)
+	tx.EncodeBinary(buf.BinWriter)
+	if buf.Err != nil {
+		return buf.Err
+	}
+	return dao.store.Put(key, buf.Bytes())
+}
+
+// IsDoubleSpend verifies that the input transactions are not double spent.
+func (dao *Dao) IsDoubleSpend(tx *transaction.Transaction) bool {
+	if len(tx.Inputs) == 0 {
+		return false
+	}
+
+	for prevHash, inputs := range tx.GroupInputsByPrevHash() {
+
+		unspent := &UnspentCoinState{}
+		key := storage.AppendPrefix(storage.STCoin, prevHash.BytesReverse())
+		if b, err := dao.store.Get(key); err == nil {
+			r := io.NewBinReaderFromBuf(b)
+			unspent.DecodeBinary(r)
+			if r.Err != nil {
+				return false
```

Comment:
Is `false` a correct behavior? May be we should return `true` or `continue` a loop?

Reply:
good question. This past was just moved. Will figure out